Add Sensoterra integration

markruys opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Sensoterra sells LoRaWAN based soil moisture probes which can be managed by a public API. We've implemented a cloud polling API which adds all Sensoterra probes the Home Assistant user owns. The integration is written and supported by Sensoterra.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- If reviewers need a Sensoterra account to test the integration, please reach out to mark.ruys@sensoterra.com.
- Brand logos and icons can be found at https://github.com/home-assistant/brands/pull/5548
- Documentation: https://github.com/home-assistant/home-assistant.io/pull/33276

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

markruys committed:
Initial version

markruys committed:
Initial version

markruys committed:
Baseline release

markruys committed:
Baseline release

joostlek commented on homeassistant/brands/sensoterra.json:
```diff
@@ -0,0 +1,5 @@
+{
```
There's only one integration for this brand, can be removed

joostlek commented on homeassistant/components/sensoterra/__init__.py:
```diff
@@ -0,0 +1,53 @@
+from .const import DOMAIN, VERSION
+from .coordinator import SensoterraCoordinator
+
+# For your initial PR, limit it to 1 platform.
```
Can be removed

joostlek commented on homeassistant/components/sensoterra/__init__.py:
```diff
@@ -0,0 +1,53 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Sensoterra from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
```
Use `entry.runtime_data` instead

joostlek commented on homeassistant/components/sensoterra/__init__.py:
```diff
@@ -0,0 +1,53 @@
+    api.set_language(hass.config.language)
+    api.set_token(entry.data["token"])
+
+    _LOGGER.info("Starting up version %s, API %s", VERSION, api.get_version())
```
Don't log on info level

joostlek commented on homeassistant/components/sensoterra/__init__.py:
```diff
@@ -0,0 +1,53 @@
+    coordinator = SensoterraCoordinator(hass, api)
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    # Fetch initial data so we have data when entities subscribe.
+    # If the refresh fails, async_config_entry_first_refresh will
+    # raise ConfigEntryNotReady and setup will try again later.
+    # If you do not want to retry setup on failure, use
+    # coordinator.async_refresh() instead.
```
Comment doesn't add much

joostlek commented on homeassistant/components/sensoterra/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+        ),
+        vol.Required(CONF_PASSWORD): TextSelector(
+            TextSelectorConfig(
+                type=TextSelectorType.PASSWORD, autocomplete="current-password"
```
Why current password

joostlek commented on homeassistant/components/sensoterra/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+
+        if user_input is not None:
+            try:
+                data = await validate_input(self.hass, user_input)
```
This can be inlined

joostlek commented on homeassistant/components/sensoterra/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+            errors=errors,
+        )
+
+    async def async_step_reconfigure(
```
Please add reauth and reconfigure in a follow up

markruys replied:
Although this is vital code, I've removed it from this PR. To be added back & reviewed in the next PR.

joostlek commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,136 @@
+        else:
+            sensors = {}
+            expiration = datetime.now(UTC) - timedelta(seconds=SCAN_INTERVAL)
+            for probe in probes:
+                for sensor in probe.sensors():
+                    if sensor.type == "MOISTURE" and sensor.unit == "SI":
+                        sensor_id = f"{sensor.id}-SI"
+                        sensor_type = "SI"
+                    else:
+                        sensor_id = sensor.id
+                        sensor_type = sensor.type
+                    sensors[sensor_id] = SensoterraSensor(
+                        expiration, probe, sensor_type, sensor
+                    )
+                if probe.battery is not None:
+                    sensor_id = f"{probe.serial}-BATTERY"
+                    sensors[sensor_id] = SensoterraSensor(expiration, probe, "BATTERY")
+                if probe.rssi is not None:
+                    sensor_id = f"{probe.serial}-RSSI"
+                    sensors[sensor_id] = SensoterraSensor(expiration, probe, "RSSI")
+                sensor_id = f"{probe.serial}-LASTSEEN"
+                sensors[sensor_id] = SensoterraSensor(expiration, probe, "LASTSEEN")
```
Together with the SemsoterraSensor object this feels like code that should be in the library rather than here

markruys replied:
I've simplified the code. The library uses a slightly different data model (a probe has a set of sensors, where HA uses a flat list of sensors instead), so some work needs to be done here.

joostlek commented on homeassistant/components/sensoterra/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "name": "Sensoterra",
+  "codeowners": ["@markruys"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/sensoterra",
+  "homekit": {},
+  "integration_type": "hub",
+  "iot_class": "cloud_polling",
+  "requirements": ["sensoterra==1.0.0"],
+  "single_config_entry": false,
+  "ssdp": [],
+  "zeroconf": []
+}
```
Please remove empty fields

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,135 @@
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+
+    def _add_devices(devices: dict[str, SensoterraSensor]) -> None:
+        sensors = [
+            SensoterraEntity(coordinator, sensor_id, sensor)
+            for sensor_id, sensor in devices.items()
+            if sensor.type
+            in ["MOISTURE", "SI", "TEMPERATURE", "BATTERY", "RSSI", "LASTSEEN"]
+        ]
+        async_add_devices(sensors)
```
```suggestion
        async_add_devices(<gen exp>)
```

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,135 @@
+    _add_devices(coordinator.data)
+
+
+class SensoterraEntity(CoordinatorEntity, SensorEntity):
```
Please type the Coordinator entity

```suggestion
class SensoterraSensor(CoordinatorEntity[..], SensorEntity):
```

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,135 @@
+        self.sensor_id = sensor_id
+        self.sensor = sensor
+
+        self._attr_has_entity_name = True
```
Can be set outside the constructor

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,135 @@
+        self._attr_has_entity_name = True
+        self._attr_unique_id = self.sensor_id
+        if sensor.type != "LASTSEEN":
+            self._attr_state_class = SensorStateClass.MEASUREMENT
+            self._attrs_native_value = float
+        self._attr_suggested_display_precision = 0
+        if sensor.soil is not None:
+            self._attr_extra_state_attributes = {
+                "soil_type": sensor.soil,
+            }
+
+        if sensor.type == "MOISTURE":
+            self._attr_device_class = SensorDeviceClass.MOISTURE
+            if sensor.depth is None:
+                self._attr_translation_key = "soil_moisture"
+            else:
+                self._attr_translation_key = "soil_moisture_at_cm"
+                self._attr_translation_placeholders = {"depth": sensor.depth}
+            self._attr_native_unit_of_measurement = PERCENTAGE
+        elif sensor.type == "SI":
+            if sensor.depth is None:
+                self._attr_name = "SI"
+            else:
+                self._attr_name = "SI @ {sensor.depth} cm"
+            self._attr_suggested_display_precision = 1
+        elif sensor.type == "TEMPERATURE":
+            self._attr_device_class = SensorDeviceClass.TEMPERATURE
+            self._attr_native_unit_of_measurement = UnitOfTemperature.CELSIUS
+        elif sensor.type == "BATTERY":
+            self._attr_device_class = SensorDeviceClass.BATTERY
+            self._attr_native_unit_of_measurement = PERCENTAGE
+            self._attr_entity_category = EntityCategory.DIAGNOSTIC
+        elif sensor.type == "RSSI":
+            self._attr_device_class = SensorDeviceClass.SIGNAL_STRENGTH
+            self._attr_native_unit_of_measurement = SIGNAL_STRENGTH_DECIBELS_MILLIWATT
+            self._attr_entity_category = EntityCategory.DIAGNOSTIC
+        elif sensor.type == "LASTSEEN":
+            self._attr_device_class = SensorDeviceClass.TIMESTAMP
+            self._attr_translation_key = "last_seen"
+            self._attr_entity_category = EntityCategory.DIAGNOSTIC
```
Please use entity descriptions

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,135 @@
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Information about the Sensoterra probe."""
+        return {
```
Please use the DeviceInfo object

markruys committed:
Refactor based on first PR feedback

markruys committed:
Refactor based on first PR feedback

zweckj commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,124 @@
+class SensoterraCoordinator(DataUpdateCoordinator):
+    """Sensoterra coordinator."""
+
+    add_devices_callback: Callable[[dict[str, SensoterraSensor]], None] | None
```
make this an instance attribute

zweckj commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,124 @@
+    async def _async_update_data(self) -> dict[str, SensoterraSensor]:
+        """Fetch data from Sensoterra Customer API endpoint."""
+        try:
+            current_sensors = set(self.async_contexts())
+            probes = await self.api.poll()
```
can both of those throw?

zweckj commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,124 @@
+            current_sensors = set(self.async_contexts())
+            probes = await self.api.poll()
+        except ApiAuthError as err:
+            raise ConfigEntryAuthFailed(err) from err
```
```suggestion
            raise ConfigEntryError(err) from err
```
until you implement reauth

zweckj commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,124 @@
+_LOGGER: logging.Logger = logging.getLogger(__package__)
+
+
+class SensoterraSensor:
```
can we move this to a models.py

zweckj commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,124 @@
+    timestamp: datetime | None = None
+    value: float | None = None
+
+    BATTERY_LEVELS: dict[str, int] = {"NORMAL": 100, "FAIR": 50, "POOR": 10}
```
make this an `Enum` or `IntEnum`

zweckj commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,124 @@
+                    SensoterraSensor(probe, sensor) for sensor in probe.sensors()
+                )
+                if probe.battery is not None:
+                    sensors.append(SensoterraSensor(probe, "BATTERY"))
```
maybe use a `StrEnum` for the type

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+SENSORS: dict[str, SensorEntityDescription] = {
+    "MOISTURE": SensorEntityDescription(
+        key="MOISTURE",
+        has_entity_name=True,
```
set this in the class, it is required to be true for new integrations anyways

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+
+    def _add_devices(devices: dict[str, SensoterraSensor]) -> None:
+        async_add_devices(
+            [
+                SensoterraEntity(coordinator, sensor)
+                for sensor in devices.values()
+                if sensor.type in SENSORS
+            ]
```
```suggestion
                SensoterraEntity(coordinator, sensor)
                for sensor in devices.values()
                if sensor.type in SENSORS
```

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+) -> None:
+    """Set up Sensoterra sensor."""
+
+    coordinator: SensoterraCoordinator = entry.runtime_data
```
```suggestion
    coordinator = entry.runtime_data
```
unnecessary if you type the entry with your custom type

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+from .const import CONFIGURATION_URL, DOMAIN
+from .coordinator import SensoterraCoordinator, SensoterraSensor
+
+SENSORS: dict[str, SensorEntityDescription] = {
```
make this a tuple and get rid of the keys, you are currently storing them twice

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+        self.sensor_id = sensor.id
+        self._attr_unique_id = sensor.id
+
+        self.entity_description = SENSORS[sensor.type]
```
normally integrations just pass in the entity description to the constructor

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+        self.entity_description = SENSORS[sensor.type]
+
+        # Make sure {depth} placeholders gets substituted.
+        self._attr_translation_placeholders = {"depth": str(sensor.depth)}
```
what happens if the sensor doesn't have a depth?

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+        if sensor.soil is not None and sensor.type in ["MOISTURE", "SI"]:
+            self._attr_extra_state_attributes = {"soil_type": sensor.soil}
+
+        self.device_info = DeviceInfo(
```
```suggestion
        self._attr_device_info = DeviceInfo(
```

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        sensors: dict[str, SensoterraSensor] = self.coordinator.data
```
```suggestion
        sensors = self.coordinator.data
```
typed automatically

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        sensors: dict[str, SensoterraSensor] = self.coordinator.data
+        for sensor in sensors.values():
+            if sensor.id == self.sensor_id:
```
the sensors is a dict, why do we need to loop and can't access the correct one directly

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+        """Initialize entity."""
+        super().__init__(coordinator, context=sensor.id)
+
+        self.sensor_id = sensor.id
+        self._attr_unique_id = sensor.id
```
why save this twice?

zweckj commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,146 @@
+
+    coordinator: SensoterraCoordinator = entry.runtime_data
+
+    def _add_devices(devices: dict[str, SensoterraSensor]) -> None:
```
```suggestion
    @callback
    def _async_add_devices(devices: dict[str, SensoterraSensor]) -> None:
```
is not blocking, so we can mark it as such

zweckj reviewed (changes requested):
you'll also need to add tests. At a minimum 100% coverage for the config flow, but I'll highly suggest you'll also add tests for the platforms (sensors can be easily done with `snapshots`)

markruys committed:
Refactoring based on second PR feedback

markruys committed:
Refactoring based on second PR feedback

markruys committed:
Merge branch 'home-assistant:dev' into sensoterra

markruys committed:
Merge branch 'sensoterra' of https://github.com/markruys/hass-core into sensoterra

zweckj commented on homeassistant/components/sensoterra/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    api.set_language(hass.config.language)
+    api.set_token(entry.data[CONF_TOKEN])
+    coordinator = SensoterraCoordinator(hass, api)
+    entry.runtime_data = coordinator
+
+    await coordinator.async_config_entry_first_refresh()
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
    
    entry.runtime_data = coordinator
```
refresh before adding it

markruys committed:
Refactoring based on PR feedback

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,149 @@
+            configuration_url=CONFIGURATION_URL,
+        )
+
+        if sensor.value is not None:
+            self._attr_native_value = sensor.value
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if self._attr_unique_id is not None:
+            sensor = self.coordinator.data[self._attr_unique_id]
+            if sensor.value is not None:
+                self._attr_native_value = sensor.value
+                self.async_write_ha_state()
```
I'd just make a `native_value` property instead, that would make this simpler

markruys replied:
Could you please elaborate on this a little bit more? It's unclear to me what you suggest to consider.

joostlek replied:
```py
@property
def native_value(self) -> StateType:
    return ....
```

markruys commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,149 @@
+        async_add_devices(
+            SensoterraEntity(coordinator, sensor, description)
+            for sensor in devices.values()
+            for description in SENSORS
+            if sensor.type == description.key
```
Well, this is what I implemented initially (https://github.com/home-assistant/core/blob/0ca493510af19733ee7f9ab98eeae0a71d9588cb/homeassistant/components/sensoterra/sensor.py), but @zweckj proposed to change it into a list of tuples because it already contained a `key` with the same value (https://github.com/home-assistant/core/pull/119642#discussion_r1641940031). What I could do is to revert back to a dictionary and construct the `key` attribute from the dictionary key. Something like 
```python
self.entity_description = description
self.entity_description.key = sensor.type
```
Would that be better?

joostlek replied:
No because entity descriptions are frozen. That little duplication of the key isn't the worst imo

markruys commented on homeassistant/components/sensoterra/config_flow.py:
```diff
@@ -0,0 +1,82 @@
+            api = CustomerApi(user_input[CONF_EMAIL], user_input[CONF_PASSWORD])
+            # We need a unique tag per HA instance
+            uuid = self.hass.data["core.uuid"]
+            expiration = datetime.now() + timedelta(TOKEN_EXPIRATION_DAYS)
```
It's a requirement of the API. We need to provide an expiration stamp.

markruys commented on homeassistant/components/sensoterra/models.py:
```diff
@@ -0,0 +1,74 @@
+    POOR = 10
+
+
+class SensoterraSensor:
```
Sensor data we receive from the API is mapped to a `SensoterraSensor` instance such that the HA integration can easily deal with it. It is a model of sensor data received from the API, hence it is defined in `models.py`.

joostlek replied:
I don't think we need this and that this adds another layer of complexity

markruys replied:
Refactored the class `SensoterraSensor` into a named tuple.

markruys commented on homeassistant/components/sensoterra/models.py:
```diff
@@ -0,0 +1,74 @@
+    LASTSEEN = auto()
+
+
+class ProbeBatteryLevel(IntEnum):
+    """Map API strings to corresponding integers."""
+
+    NORMAL = 100
+    FAIR = 50
+    POOR = 10
```
"NORMAL", "FAIR", or "POOR"

joostlek replied:
Let's just make that an enum sensor instead of translating them to 3 integers

markruys replied:
I've moved the translation from battery level enum to percentage to the PyPI implementation

joostlek replied:
That still doesn't make sense. If we get a list of only those 3 options back we should keep them like that imo

markruys commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,149 @@
+        native_unit_of_measurement=SIGNAL_STRENGTH_DECIBELS_MILLIWATT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    SensorEntityDescription(
+        key=ProbeSensorType.LASTSEEN,
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        translation_key="last_seen",
+    ),
```
I've removed the last_seen sensor

markruys committed:
Refactoring based on PR feedback

markruys committed:
Remove extra attribute soil type

Soil type isn't really a sensor, but more like a configuration entity.
Move soil type to a different PR to keep this PR simpler.

markruys commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,149 @@
+            ProbeSensorType.MOISTURE,
+            ProbeSensorType.SI,
+        ]:
+            self._attr_extra_state_attributes = {"soil_type": sensor.soil}
```
Actually, soil type isn't really a sensor, but more like a configuration entity. I've removed the soil type extra attribute to keep this PR simpler. I'll create a separate PR for soil type.

markruys committed:
Refactor SensoterraSensor to a named tuple

joostlek commented on homeassistant/components/sensoterra/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+                    f"Home Assistant {uuid}", "READONLY", expiration
+                )
+            except StInvalidAuth as exp:
+                LOGGER.error(
+                    "Login attempt with %s: %s", user_input[CONF_EMAIL], exp.message
+                )
+                errors["base"] = "invalid_auth"
+            except StTimeout:
+                LOGGER.error("Login attempt with %s: time out", user_input[CONF_EMAIL])
+                errors["base"] = "cannot_connect"
```
not sure what the logs add

markruys replied:
They show a reason why the authentication attempt failed. Can be handy to understand why the integration is not working. I propose to leave it there, but if you like I'll remove it.

joostlek commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,68 @@
+            update_interval=timedelta(minutes=SCAN_INTERVAL_MINUTES),
+        )
+        self.api = api
+        self.add_devices_callback: Callable[[list[SensoterraSensor]], None] | None = (
+            None
+        )
```
You can just use the normal listener setup from the coordinator

markruys replied:
I'm a bit lost here, could you please clarify a bit? 

The idea is that the API gives a list of probes with sensors, and that we might need to add newly registered sensors to HA.

joostlek commented on homeassistant/components/sensoterra/coordinator.py:
```diff
@@ -0,0 +1,68 @@
+                ]
+            )
+
+        return [
+            SensoterraSensor(probe, sensor)
+            for probe in probes
+            for sensor in probe.sensors()
+        ]
```
Why don't we return a structure more like:
```py
@dataclass
class SensoterraData:
    probes: dict[str, Probe]
    sensors: dict[str, Sensor]
```
with the `str` being the id of both types. This leaves you with easy access to both values without having to looping over them.

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,139 @@
+            "depth": "?" if sensor.depth is None else str(sensor.depth)
+        }
+
+        self.entity_description = SENSORS[ProbeSensorType[sensor.type]]
```
Let's make the entity_description a parameter instead

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,139 @@
+        super().__init__(coordinator, context=sensor.id)
+
+        self._attr_unique_id = sensor.id
+        self._attr_translation_placeholders = {
+            "depth": "?" if sensor.depth is None else str(sensor.depth)
+        }
```
I would rather see a separate sensor telling the depth

markruys replied:
Depth is not something you measure, it's a physical propery of the soul moisture sensor: the length of the pins hammered into the soil. Some probes even have six soil moisture sensors at 6 different fixed depths. For the user it's important to know which sensor at which depth the soil moisture sensor is reporting. That's why depth is part of the sensor name.

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,139 @@
+    @property
+    def native_value(self) -> StateType:
+        """Return the sensor value reported by the API."""
+        sensor = self.coordinator.get_sensor(self._attr_unique_id)
+        if sensor is None:
+            return None
```
It should be made unavailable instead. use the `available` property

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,139 @@
+        if sensor.timestamp is None or sensor.timestamp < expiration:
+            return None
+
+        return sensor.value  # type: ignore[no-any-return]
```
The type ignore is needed because you use `.strict-typing` but your library is either weakly typed or the library does not have a `py.typed` file, thus mypy can't read the fields. We strongly recommend to fix this since it greatly improves code quality.

markruys committed:
Implement feedback on PR

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,152 @@
+            configuration_url=CONFIGURATION_URL,
+        )
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        sensor = self.coordinator.get_sensor(self._attr_unique_id)
+        if sensor is None:
+            self._attr_native_value = None
+            return False
+
+        self._attr_native_value = sensor.value
+
+        if sensor.timestamp is None:
+            return False
+
+        # Expire sensor if no update within the last few days.
+        expiration = datetime.now(UTC) - timedelta(days=SENSOR_EXPIRATION_DAYS)
+        return sensor.timestamp >= expiration
```
This is a property, this should not have side effects. You should split this up into 2 properties. One for native value and one for available.

If you feel extra fancy, you can add a property for `sensor` and then you can easily access `self.sensor` to avoid having to write `self.coordinator.get_sensor(...)` in both properties

joostlek commented on homeassistant/components/sensoterra/sensor.py:
```diff
@@ -0,0 +1,152 @@
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        sensor = self.coordinator.get_sensor(self._attr_unique_id)
```
I would propose to also set the sensor id to something like `self._sensor_id`, so we can avoid having to use shorthands as parameters (even though it works, it always feels dirty to use)

joostlek commented on homeassistant/components/sensoterra/strings.json:
```diff
@@ -0,0 +1,41 @@
+        }
+      },
+      "reconfigure": {
+        "description": "Enter credentials to obtain a new token",
```
```suggestion
        "description": "[%key:component::sensoterra::config::step::user::description%]",
```

joostlek commented on homeassistant/components/sensoterra/strings.json:
```diff
@@ -0,0 +1,41 @@
+      "si_at_cm": {
+        "name": "SI @ {depth} cm"
+      },
+      "last_seen": {
+        "name": "Last seen"
+      }
```
unused

joostlek commented on homeassistant/components/sensoterra/models.py:
```diff
@@ -0,0 +1,13 @@
+from enum import StrEnum, auto
+
+
+class ProbeSensorType(StrEnum):
```
Move this to `sensor.py`. You only use every type once, so maybe its time to ask ourselves how much this adds

joostlek commented on tests/components/sensoterra/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+async def test_form(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
+    """Test we get the form."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

joostlek commented on tests/components/sensoterra/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```
Please also change this at the rest

joostlek commented on tests/components/sensoterra/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+
+    hass.data["core.uuid"] = HASS_UUID
+
+    with patch(
+        "sensoterra.customerapi.CustomerApi.get_token",
+        return_value=API_ACCESS_TOKEN,
+    ):
```
Let's create a global mock so we can avoid having to patch this inline every time. Checkout `airgradient` for an example

joostlek commented on tests/components/sensoterra/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+                CONF_PASSWORD: API_PASSWORD,
+            },
+        )
+        await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on homeassistant/components/sensoterra/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+)
+
+
+class SensoterraConfigFlow(ConfigFlow, domain=DOMAIN):
```
Do we have something unique that we can use as unique identifier for the config flow to avoid the user to set up twice

markruys replied:
No, I don't think so. The API token gives access to possibly zero or more devices. The API token is not stable, nor is the email address of the customer.

joostlek replied:
What kind of token do we get? JWT?

joostlek replied:
Otherwise we have to do `self._async_abort_entries_match` to make sure we can't add duplicates

markruys replied:
As I also maintain the Sensoterra API, I have change the API token from a hash to a JWT token. This token contains the unique ID we need. I've implemented `await self.async_set_unique_id(device_unique_id)`.

joostlek commented on tests/components/sensoterra/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+    hass.data["core.uuid"] = HASS_UUID
+
+    with patch(
+        "sensoterra.customerapi.CustomerApi.get_token",
+        side_effect=StTimeout,
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_EMAIL: API_EMAIL,
+                CONF_PASSWORD: API_PASSWORD,
+            },
+        )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {"base": "cannot_connect"}
+
+    # Make sure the config flow tests finish with either an
+    # FlowResultType.CREATE_ENTRY or FlowResultType.ABORT so
+    # we can show the config flow is able to recover from an error.
+    with patch(
+        "sensoterra.customerapi.CustomerApi.get_token",
+        return_value=API_ACCESS_TOKEN,
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_EMAIL: API_EMAIL,
+                CONF_PASSWORD: API_PASSWORD,
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
+    assert result["title"] == API_EMAIL
+    assert result["data"] == {
+        CONF_EMAIL: API_EMAIL,
+        CONF_TOKEN: API_ACCESS_TOKEN,
+    }
+    assert len(mock_setup_entry.mock_calls) == 1
```
These 2 tests can be combined using `pytest.mark.parametrize` if you do a global search in the code you can find examples of this for the config flow exception tests

markruys committed:
Remove .coveragerc

markruys committed:
Merge branch 'dev' into sensoterra

joostlek commented:
https://github.com/home-assistant/core/pull/119642#discussion_r1744310999

markruys committed:
Add async_set_unique_id to config flow

joostlek commented on homeassistant/components/sensoterra/config_flow.py:
```diff
@@ -53,6 +54,13 @@ async def async_step_user(
+                decoded_token = decode(
+                    token, algorithm="HS256", options={"verify_signature": False}
+                )
+                device_unique_id = decoded_token["sub"]
+                await self.async_set_unique_id(device_unique_id)
+                self._abort_if_unique_id_configured()
```
can we move these 3 lines to the `else`?

joostlek commented on tests/components/sensoterra/conftest.py:
```diff
@@ -0,0 +1,24 @@
+from .const import API_TOKEN
+
+
+@pytest.fixture
+def mock_setup_entry() -> Generator[AsyncMock, None, None]:
+    """Override async_setup_entry."""
+    with (
+        patch(
+            "homeassistant.components.sensoterra.async_setup_entry",
+            return_value=True,
+        ),
+        patch(
+            "sensoterra.customerapi.CustomerApi.get_token",
+            return_value=API_TOKEN,
+        ) as mock_setup_entry,
+    ):
+        yield mock_setup_entry
```
Please split the mocks

joostlek replied:
Oh and we should patch the CustomerApi object where we use it, so `homeassistant.components.sensoterra.*`

markruys replied:
I just pushed code which test the `unique_id`. 

> Oh and we should patch the CustomerApi object where we use it, so homeassistant.components.sensoterra.*

I can't patch like: 
```python
@pytest.fixture
def mock_get_token() -> Generator[AsyncMock, None, None]:
    """Override async_setup_entry."""
    with (
        patch(
            "sensoterra.customerapi.*",
            return_value=API_TOKEN,
        ) as mock_entry,
    ):
        yield mock_entry
```
It gives me a ` AttributeError: <module 'sensoterra.customerapi' from '.../customerapi.py'> does not have the attribute '*'`. In the config flow tests we only need `get_token()`, so I guess we're good for now.

I splitted the patches above, but I eventually remove the patch on `homeassistant.components.sensoterra.async_setup_entry`. Tests do not seem to be affected.

joostlek replied:
No I mean, `homeassistant.components.sensoterra.CustomerApi` will patch the usage of `CustomerApi` in `__init__.py`. Check Airgradient for an example

joostlek commented on tests/components/sensoterra/test_config_flow.py:
```diff
@@ -0,0 +1,120 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+@pytest.mark.parametrize("test", [Test.INVALID_AUTH, Test.TIMEOUT, Test.DECODE_ERROR])
```
Rather just add the exceptions you want to put in `side_effect` as parameter. We don't want if statements in the tests

markruys replied:
Yes, indeed, that was poorly written. Fixed.

markruys committed:
Small fix based on feedback

joostlek reviewed (changes requested):
I am missing the test where we add a mock_config_entry with the same unique_id, and then try to setup the account again and test the flow aborts

markruys committed:
Add test form unique_id

joostlek committed:
Fix

joostlek committed:
Fix
